Merlin_Fork_Options.txt
Merlin Fork Custom Options

----------------------------------------------------------------------------------------------------------------
Stop copy of syslog files to JFFS

        Can be set in gui:      Yes (as of Update-11, Administration/System/Save syslog to JFFS)
        Enable via NVRAM:       nvram set jffs2_log=1
                                nvram commit && reboot

        Routers Supported:      AC66,AC56,AC68
        Released:               Update-02
        Reboot Required:        Yes
        Notes:                  Existing logs must be manually deleted from /jffs (syslog.log and syslog.log-1)
----------------------------------------------------------------------------------------------------------------
Enable Comcast DSCP fix

        Can be set in gui:      Yes (Firewall/General/Enable DSCP override)
        Enable via NVRAM:       nvram set DSCP_fix_enable=1
                                nvram commit && reboot

        Routers Supported:      All
        Released:               Update-07
        Reboot Required:        Yes
        Notes:                  For SOME Comcast service areas it has been seen that the
                                Differentiated Services Code Point (DSCP) value is incorrectly set to a low
                                priority, resulting in decreased internet throughput for wireless IPv4 transfers.
                                This fix overrides that incorrect value.
----------------------------------------------------------------------------------------------------------------
Display Traffic Monitor graphs in
Mb/s (default is KB/s)

        Can be set in gui:      Yes (Tools/Other Settings/Traffic Monitor graph units)
        Enable via NVRAM:       nvram set rstats_units=1
                                nvram commit

        Routers Supported:      All
        Released:               Update-07
        Reboot Required:        No
        Notes:                  Refresh or open the TM page in the browser
----------------------------------------------------------------------------------------------------------------
Multiple listening addresses and
ports for SSH access

        Can be set in gui:      No
        Enable via NVRAM:       nvram set sshd_addr=[address:]port,[address:]port,[address:]port,[address:]port
                                nvram commit && reboot

        Routers Supported:      All
        Released:               Update-07
        Reboot Required:        Yes
        Notes:                  Up to 4 address:port or port values may be specified separated by commas
                                This will override the port setting specified in the gui
                                This modifies a previous commit in Update-05 (was a single address only and now
                                        need to specify a port when specifying an address)
----------------------------------------------------------------------------------------------------------------
Enable full syslog records for
remote monitoring

        Can be set in gui:      Yes (Update-15)
        Enable via NVRAM:       nvram set log_small=0
                                nvram commit && reboot

        Routers Supported:      All
        Released:               Update-08 (NVRAM) Update-15 (GUI)
        Reboot Required:        Yes
        Notes:                  This will include the full header (including hostname and priority level) in the
                                        syslogd output to enable remote monitoring
----------------------------------------------------------------------------------------------------------------
Save HTTPS certificate

        Can be set in gui:      Yes (Administration/System/Save HTTPS certificate)
        Enable via NVRAM:       nvram set https_crt_save=1
                                nvram commit && reboot

        Routers Supported:      All
        Released:               Update-11
        Reboot Required:        Yes
        Notes:                  If you are using HTTPS validation, the security certificate is regenerated at every
                                  boot, requiring you to install a new certificate in your browser.  This
                                  option saves the certificate across reboots negating the need to reinstall
                                  in your browser.
----------------------------------------------------------------------------------------------------------------
Filter IPv6 Neighbor Solicitation

        Can be set in gui:      Yes (Firewall/General/Filter IPv6 Neighbor Solicitation)
        Enable via NVRAM:       ipv6_neighsol_drop=1 (1=Yes, 0=No)
                                nvram commit
                                service restart_firewall

        Routers Supported:      All
        Released:               Update-11
        Reboot Required:        No
        Notes:                  Control iptables rule dropping IPv6 neighbor solicitation requests
----------------------------------------------------------------------------------------------------------------
CROND logging options

        Can be set in gui:      No
        Enable via NVRAM:       nvram set cron_loglevel=N (default is 8)
                                nvram commit && reboot

        Routers Supported:      All
        Released:               Update-11
        Reboot Required:        Yes
        Notes:                  Setting cron_loglevel=N sets the logging level to N
----------------------------------------------------------------------------------------------------------------
Scan media files at every boot

        Can be set in gui:      Yes (USB Applications/Servers Center/Media Server/Scan media files at every boot)
        Enable via NVRAM:       nvram set dms_scan=1
                                nvram commit

        Routers Supported:      All
        Released:               Update-12
        Reboot Required:        No
        Notes:                  Setting to No will only scan media files when a change is detected
----------------------------------------------------------------------------------------------------------------
Enable IPv6 MTU Advertisement

        Can be set in gui:      Yes (IPv6/Enable IPv6 MTU Advertisement)
        Enable via NVRAM:       nvram set ipv6_radvd_mtu=1 (0 to disable)
                                nvram commit

        Routers Supported:      All
        Released:               Update-13
        Reboot Required:        No
        Notes:                  Default is to enable advertisement for IPv6 on all connection types
----------------------------------------------------------------------------------------------------------------
Advertise Router as IPv6 DNS Server

        Can be set in gui:      Yes (IPv6/Advertise router as DNS server)
        Enable via NVRAM:       nvram set ipv6_dns_router=1
                                nvram commit

        Routers Supported:      All
        Released:               Update-14
        Reboot Required:        Yes
        Notes:                  Enable IPv6 DNS lookups to be cached by dnsmasq
----------------------------------------------------------------------------------------------------------------
Stealth Mode - Show power status only

        Can be set in gui:      Yes (Tools/Other Settings/Stealth Mode/Show power status only)
        Enable via NVRAM:       nvram set led_disable=2
                                nvram commit
                                service restart_leds

        Routers Supported:      All
        Released:               Update-14
        Reboot Required:        No
        Notes:                  Show only the power status led
----------------------------------------------------------------------------------------------------------------
HTTP Lan Port

        Can be set in gui:      Yes (Administration/System/HTTP Lan Port)
        Enable via NVRAM:       No, nvram variable http_lanport must be set thru gui

        Routers Supported:      All
        Released:               Update-14
        Reboot Required:        Yes
        Notes:                  Changes HTTP port for gui access from default port 80
                                Reset to default of 80 before loading firmware which does not support
                                  changing of the port
----------------------------------------------------------------------------------------------------------------
QOS SFQ limit

        Can be set in gui:      No
        Enable via NVRAM:       nvram set qos_sfql=[0,off-default  1,auto  2-127,value]
                                nvram commit

        Routers Supported:      All
        Released:               Update-14
        Reboot Required:        Yes
        Notes:                  Changes SFQ qdisc packet limit
                                A value of 1 automatically selects a value based on upload speed (recommended)
                                A value of 2-127 sets this value as the packet limit
----------------------------------------------------------------------------------------------------------------
SNTP Server

        Can be set in gui:      Yes (Administration/System/Enable router as local SNTP server)
        Enable via NVRAM:       nvram set ntpd_server=1
                                nvram commit
                                service restart_time

        Routers Supported:      All
        Released:               Update-15
        Reboot Required:        No
        Notes:                  Enables the router to act as an SNTP time server
                                User 'router.asus.com' as your SNTP server
----------------------------------------------------------------------------------------------------------------
Networkmap UNIX/Linux Clients as PC

        Can be set in gui:      No setting required
        Enable via NVRAM:       No setting required

        Routers Supported:      All
        Released:               Update-15
        Reboot Required:        No
        Notes:                  You must install SAMBA and winbind on the client for the client to be
                                  recognized as a PC
----------------------------------------------------------------------------------------------------------------
Disable Avahi (mDNS) Reflector

        Can be set in gui:      No
        Enable via NVRAM:       nvram set mdns_reflector=0
                                nvram commit
                                service restart_mdns

        Routers Supported:      All
        Released:               Update-15
        Reboot Required:        No
        Notes:                  Changes the default state of the Avahi reflector to disabled
----------------------------------------------------------------------------------------------------------------
Enable logging of successful NTP updates

        Can be set in gui:      No
        Enable via NVRAM:       nvram set ntp_log=0
                                nvram commit

        Routers Supported:      All
        Released:               Update-15
        Reboot Required:        No
        Notes:                  Successful NTP updates are not logged by default, fails are always logged
                                This option enables logging of successful updates
----------------------------------------------------------------------------------------------------------------
Set NTP update interval

        Can be set in gui:      Yes (Administration/System/Router NTP update interval (hours))
        Enable via NVRAM:       nvram set ntp_update=[interval in hours]
                                nvram commit

        Routers Supported:      All
        Released:               Update-15 w/o gui support, Update-16 w/ gui support
        Reboot Required:        No
        Notes:                  Default is that the router will update with its NTP server every hour
                                This option changes the number of hours between updates.  The value is not currently error
                                  checked and should be an integer between 1 and 24 (hours)
                                A value of '0' disables NTP updates and the time must be set manually (Update-16)
----------------------------------------------------------------------------------------------------------------
Set NTP alternate time server

        Can be set in gui:      Yes (Administration/System/NTP Server (Alternate))
        Enable via NVRAM:       nvram set ntp_update=[interval in hours]
                                nvram commit

        Routers Supported:      All
        Released:               Update-15 w/o gui support, Update-16 w/ gui support
        Reboot Required:        No
        Notes:                  Default is that the router will update with its NTP server every hour
                                This option changes the number of hours between updates.  The value is not currently error
                                  checked and should be an integer between 1 and 24 (hours)
                                A value of '0' disables NTP updates and the time must be set manually (Update-16)

----------------------------------------------------------------------------------------------------------------
Set number of threads for NVS server

        Can be set in gui:      No
        Enable via NVRAM:       nvram set nfsd_threads=x (where x is the number of threads)
                                nvram commit & service restart_nfs

        Routers Supported:      All
        Released:               Update-17
        Reboot Required:        No
        Notes:                  The default number of threads has now been raised to two (instead of a single thread)
---------------------------------------------------------------------------------------------------------------
Set preferred interface for bandwidth limiter

        Can be set in gui:      No
        Enable via NVRAM:       nvram set qos_iface=[iface]
                                nvram commit && reboot

        Routers Supported:      All
        Released:               Update-17
        Reboot Required:        No
        Notes:                  The default interface is eth0 (standard WAN interface).   This may be useful if running
                                  an OpenVPN client and wish to limit that bandwidth (set to tun11 for example for
                                  vpnclient1.
---------------------------------------------------------------------------------------------------------------
Use alternate NAT Loopback implementation (ASUS)

        Can be set in gui:      No
        Enable via NVRAM:       nvram set fw_nat_loopback=1
                                nvram commit && reboot

        Routers Supported:      All
        Released:               Update-17
        Reboot Required:        No
        Notes:                  The default NAT Loopback is that developed by Merlin and has been used in all fork
                                  firmware levels which have been released.   Merlin's implementation employs
                                  iptables marks in it's implementation.   This alternate method does not use
                                  marks and may be helpful if the user is employing custom marks.   Note also
                                  that in V17 the standard Merlin implementation has been changed to only use
                                  a single bit, to aid users who what to perform custom packet marking.
---------------------------------------------------------------------------------------------------------------
Ability to disable DNSSEC permissive mode

        Can be set in gui:      No
        Enable via NVRAM:       nvram set dnssec_check_unsigned=1
                                nvram commit && reboot

        Routers Supported:      All
        Released:               Update-17
        Reboot Required:        No
        Notes:                  By default, DNSSEC will also accept unsigned responses.  This option forces DNSSEC
                                  to only accept signed responses.  USE WITH CAUTION:  This can disable dns lookups
                                  if no DNSSEC enabled servers are available.
---------------------------------------------------------------------------------------------------------------
Ability to enable DFS channels for N66U/AC66U-MIPS EU routers

        Can be set in gui:      No
        Enable via NVRAM:       nvram set wl_dfs_enable=1
                                nvram commit && reboot

        Routers Supported:      N66U/R/W, AC66U_B0
        Released:               Update-18
        Reboot Required:        Yes
        Notes:                  The base 374.43 firmware only supports the lower channels on the 5GHz band in
                                  EU configured routers.
                                This setting enables support for the DFS and upper 5GHz channels with full regulatory

Changelog.txt

Source:  https://github.com/john9527/asuswrt-merlin : branch 374.43_2-update
NOTE: [*] Indicated fork unique change

Changelog
---------
374.43_52E7j9527  (19-December-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: nettle: update to 3.7.3
* CHANGED: curl: update to 7.79.1
* CHANGED: make: enable additional ipv6 conntrack modules (MIPS)
* CHANGED: kernel: enable additional ipv6 modules (MIPS)
* CHANGED: qos: cannot support ipv6 qos on mips
* FIXED: firewall: initialize all protection server rule chains
* FIXED: shared: unlink parent leases if ext data option disabled
* FIXED: iptables: mips: fix output formatting of timestart/timestop rules
* FIXED: repo: add repo missing files
* FIXED: httpd: Enable POST/GET handling for **.js


374.43_52E3j9527  (20-October-2021) ARM SDK, MIPS,ARM KRACK
* FIXED: webui: fix ipv6 setup page error


374.43_52E2j9527  (18-October-2021) ARM SDK, MIPS,ARM KRACK
* NEW: webui: support openwrt-orion for enhanced ap data (webui)
* CHANGED: Revert "usb: wsdd: allow multiple listening interfaces via nvram"
* CHANGED: rndis_host: support Novatel Verizon USB730L (MIPS)
* CHANGED: rndis_host: support Novatel Verizon USB730L (ARM)
* CHANGED: webui: update jquery.xdomainajax.js to 1.0.0
* CHANGED: openvpn: update to 2.5.4
* FIXED: stubby: update open logfile with append; don't redirect stdout (reapply)


374.43_51D7j9527  (25-September-2021) ARM SDK, MIPS,ARM KRACK
* NEW: shared: support openwrt for enhanced ap data (function)
* CHANGED: dnsmasq: update to 2.86
* CHANGED: watchdog: change http check disable covers both http/httpd
* CHANGED: usb: wsdd: allow multiple listening interfaces via nvram
* FIXED: webui: remove orphan call to deprecated wan uptime function
* FIXED: webui: openvpn: set client off flag when resetting client


374.43_51D6j9527  (12-September-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: samba: only enable dns proxy w/wins if dnsmasq started
* CHANGED: webui: remove auto redirect to static routes from DHCP server page
* CHANGED: webui: webui: reboot all except ac68U when disabling native ipv6
* CHANGED: getdns: revert upstream Record and guard UDP max payload size with servers 6cb1593
* FIXED: services: dnsmasq: smbd_wins sets lan ipaddr insted of wildcard addr
* FIXED: services: dnsmasq: smbd_wins not used if url filter active


374.43_51D4j9527  (09-September-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: webui: add javascript jquery to wireless status
* CHANGED: webui: (backport) update dst start/stop data
* CHANGED: webui: fix hang when changing static routes


374.43_51E3j9527  (28-August-2021) ARM SDK, MIPS,ARM KRACK 
* Equivalent to 51D3j9527
* CHANGED: Revert "nvram: nvram_get return static ptr instead of mmap ptr"


374.43_51D2j9527  (26-August-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: openssl11: update to 1.1.1l
* CHANGED: getdns/stubby: update to 1.7.0/0/4.0
* CHANGED: stubby fork customizations
* CHANGED: curl: update to 7.78.0


374.43_50DAj9527  (11-August-2021) ARM SDK, MIPS,ARM KRACK
* FIXED: webui: add jQuery.noConflict to wireless status page
* CHANGED: watchdog: shorten timeout for httpd_check
* CHANGED: ncurses: rename directory to sync with merlinng
* CHANGED: make: update nano/ncurses build recipes
* CHANGED: watchdog: remove FUNCTION refs from wireless/reboot syslogs
* CHANGED: Revert "make: remove bwdpi override for nf_conntrack_events"


374.43_50D9j9527  (1-August-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: watchdog: add nvram switch to disable auto restart


374.43_50E8j9527  (21-July-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: webui: improve method of getting dnsmasq leases in ap mode
* CHANGED: webui: get parent leases in ap mode via gui instead of inline
* CHANGED: networkmap: check local hosts file to resolve hostnames


374.43_50D7j9527  (10-July-2021) ARM SDK, MIPS,ARM KRACK
* NEW: webui: add tools option for get parent data in ap mode
* CHANGED: networkmap: get parent leases file when in ap mode
* CHANGED: shared: clear parent leases file before refresh
* CHANGED: avahi: (upstream) cleanup: ensure entries are dead for at least 1s
* CHANGED: webui: allow 64 chars for url filter
* CHANGED: dnsmasq: urlfilter: implement dns blocking for url filter
* CHANGED: firewall: urlfilter: add ipv6 support for url filter


374.43_50D3j9527  (25-June-2021) ARM SDK, MIPS,ARM KRACK
* NEW: mdns: samba: add option to broadcast sharing via avahi (arm only)
* NEW: webui: retrieve parent leases for wireless log in ap mode
* CHANGED: avahi: add required nogroup to group system file
* CHANGED: services: update start/stop_mdns in service functions
* CHANGED: openvpn: update to 2.5.3
* FIXED: wireless: show ap isolate status/link when in ap mode


374.43_50D1j9527  (19-June-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: watchdog: reboot scheduler set SIGTERM instead of linux reboot
* CHANGED: watchdog: add syslog for reboot scheduler events
* CHANGED: wireless: add syslog for radio scheduler events
* CHANGED: watchdog: ensure watchdog shut down during reboot
* CHANGED: ntp: optimize restart of dns services
* CHANGED: services: stubby: do not restart during reboot
* CHANGED: services: dnsmasq: check for reboot earlier during start
* CHANGED: services: remove unnecessary stop_dnscrypt/stubby from service functions
* FIXED: init: pass reboot global to sighandler via nvram


374.43_49D6j9527  (07-June-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: wl_arm: fragattack patches to wireless driver (ac68u)
* CHANGED: smb: make config compatible with avahi smb.service


374.43_49D5j9527  (28-May-2021) ARM SDK, MIPS,ARM KRACK
* CHANGED: webui: update chart.min.js to 2.9.4
* CHANGED: webui: (backport) allow OpenVPN CommonName up to 64 chars long for server clientlist
* CHANGED: nano: update to 5.7
* CHANGED: curl: update to 7.76.1
* CHANGED: webui: update ASUS ouiDB; add missing TP-LINK entries
* CHANGED: webui: cache ouiDB.js for faster response
* CHANGED: avahi: improve generated config
* FIXED: dnsmasq: (upstream) fix bug in TCP process handling



release/src/router/dnsmasq/src/config.h
/* dnsmasq is Copyright (c) 2000-2021 Simon Kelley

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; version 2 dated June, 1991, or
   (at your option) version 3 dated 29 June, 2007.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#define FTABSIZ 150 /* max number of outstanding requests (default) */
#define MAX_PROCS 20 /* max no children for TCP requests */
#define CHILD_LIFETIME 150 /* secs 'till terminated (RFC1035 suggests > 120s) */
#define TCP_MAX_QUERIES 100 /* Maximum number of queries per incoming TCP connection */
#define TCP_BACKLOG 32  /* kernel backlog limit for TCP connections */
#define EDNS_PKTSZ 4096 /* default max EDNS.0 UDP packet from RFC5625 */
#define SAFE_PKTSZ 1280 /* "go anywhere" UDP packet size */
#define KEYBLOCK_LEN 40 /* choose to minimise fragmentation when storing DNSSEC keys */
#define DNSSEC_WORK 50 /* Max number of queries to validate one question */
#define TIMEOUT 10     /* drop UDP queries after TIMEOUT seconds */
#define FORWARD_TEST 50 /* try all servers every 50 queries */
#define FORWARD_TIME 20 /* or 20 seconds */
#define UDP_TEST_TIME 60 /* How often to reset our idea of max packet size. */
#define SERVERS_LOGGED 30 /* Only log this many servers when logging state */
#define LOCALS_LOGGED 8 /* Only log this many local addresses when logging state */
#define LEASE_RETRY 60 /* on error, retry writing leasefile after LEASE_RETRY seconds */
#define CACHESIZ 150 /* default cache size */
#define TTL_FLOOR_LIMIT 3600 /* don't allow --min-cache-ttl to raise TTL above this under any circumstances */
#define MAXLEASES 1000 /* maximum number of DHCP leases */
#define PING_WAIT 3 /* wait for ping address-in-use test */
#define PING_CACHE_TIME 30 /* Ping test assumed to be valid this long. */
#define DECLINE_BACKOFF 600 /* disable DECLINEd static addresses for this long */
#define DHCP_PACKET_MAX 16384 /* hard limit on DHCP packet size */
#define SMALLDNAME 50 /* most domain names are smaller than this */
#define CNAME_CHAIN 10 /* chains longer than this atr dropped for loop protection */
#define DNSSEC_MIN_TTL 60 /* DNSKEY and DS records in cache last at least this long */
#define HOSTSFILE "/etc/hosts"
#define ETHERSFILE "/etc/ethers"
#define DEFLEASE 3600 /* default DHCPv4 lease time, one hour */
#define DEFLEASE6 (3600*24) /* default lease time for DHCPv6. One day. */
#define CHUSER "nobody"
#define CHGRP "dip"
#define TFTP_MAX_CONNECTIONS 50 /* max simultaneous connections */
#define LOG_MAX 5 /* log-queue length */
#define RANDFILE "/dev/urandom"
#define DNSMASQ_SERVICE "uk.org.thekelleys.dnsmasq" /* Default - may be overridden by config */
#define DNSMASQ_PATH "/uk/org/thekelleys/dnsmasq"
#define DNSMASQ_UBUS_NAME "dnsmasq" /* Default - may be overridden by config */
#define AUTH_TTL 600 /* default TTL for auth DNS */
#define SOA_REFRESH 1200 /* SOA refresh default */
#define SOA_RETRY 180 /* SOA retry default */
#define SOA_EXPIRY 1209600 /* SOA expiry default */
#define LOOP_TEST_DOMAIN "test" /* domain for loop testing, "test" is reserved by RFC 2606 and won't therefore clash */
#define LOOP_TEST_TYPE T_TXT

/* compile-time options: uncomment below to enable or do eg.
   make COPTS=-DHAVE_BROKEN_RTC

HAVE_BROKEN_RTC
   define this on embedded systems which don't have an RTC
   which keeps time over reboots. Causes dnsmasq to use uptime
   for timing, and keep lease lengths rather than expiry times
   in its leases file. This also make dnsmasq "flash disk friendly".
   Normally, dnsmasq tries very hard to keep the on-disk leases file
   up-to-date: rewriting it after every renewal.  When HAVE_BROKEN_RTC
   is in effect, the lease file is only written when a new lease is
   created, or an old one destroyed. (Because those are the only times
   it changes.) This vastly reduces the number of file writes, and makes
   it viable to keep the lease file on a flash filesystem.
   NOTE: when enabling or disabling this, be sure to delete any old
   leases file, otherwise dnsmasq may get very confused.

HAVE_LEASEFILE_EXPIRE
   define this if you want to enable lease file update with expire
   timeouts instead of expiry times or lease lengths, if HAVE_BROKEN_RTC
   is also enabled. Lease file will be rewritten upon SIGUSR2 signal
   reception and/or dnsmasq termination.

HAVE_TFTP
   define this to get dnsmasq's built-in TFTP server.

HAVE_DHCP
   define this to get dnsmasq's DHCPv4 server.

HAVE_DHCP6
   define this to get dnsmasq's DHCPv6 server. (implies HAVE_DHCP).

HAVE_SCRIPT
   define this to get the ability to call scripts on lease-change.

HAVE_LUASCRIPT
   define this to get the ability to call Lua script on lease-change. (implies HAVE_SCRIPT)

HAVE_DBUS
   define this if you want to link against libdbus, and have dnsmasq
   support some methods to allow (re)configuration of the upstream DNS
   servers via DBus.

HAVE_UBUS
   define this if you want to link against libubus

HAVE_IDN
   define this if you want international domain name 2003 support.

HAVE_LIBIDN2
   define this if you want international domain name 2008 support.

HAVE_CONNTRACK
   define this to include code which propagates conntrack marks from
   incoming DNS queries to the corresponding upstream queries. This adds
   a build-dependency on libnetfilter_conntrack, but the resulting binary will
   still run happily on a kernel without conntrack support.

HAVE_IPSET
    define this to include the ability to selectively add resolved ip addresses
    to given ipsets.

HAVE_AUTH
   define this to include the facility to act as an authoritative DNS
   server for one or more zones.

HAVE_CRYPTOHASH
   include just hash function from crypto library, but no DNSSEC.

HAVE_DNSSEC
   include DNSSEC validator.

HAVE_DUMPFILE
   include code to dump packets to a libpcap-format file for debugging.

HAVE_LOOP
   include functionality to probe for and remove DNS forwarding loops.

HAVE_INOTIFY
   use the Linux inotify facility to efficiently re-read configuration files.

NO_ID
   Don't report *.bind CHAOS info to clients, forward such requests upstream instead.
NO_TFTP
NO_DHCP
NO_DHCP6
NO_SCRIPT
NO_LARGEFILE
NO_AUTH
NO_DUMPFILE
NO_LOOP
NO_INOTIFY
   these are available to explicitly disable compile time options which would 
   otherwise be enabled automatically or which are enabled  by default 
   in the distributed source tree. Building dnsmasq
   with something like "make COPTS=-DNO_SCRIPT" will do the trick.
NO_GMP
   Don't use and link against libgmp, Useful if nettle is built with --enable-mini-gmp.

LEASEFILE
CONFFILE
RESOLVFILE
   the default locations of these files are determined below, but may be overridden
   in a build command line using COPTS.

*/

/* Defining this builds a binary which handles time differently and works better on a system without a
   stable RTC (it uses uptime, not epoch time) and writes the DHCP leases file less often to avoid flash wear.
*/

/* #define HAVE_BROKEN_RTC */
/* #define HAVE_LEASEFILE_EXPIRE */

/* The default set of options to build. Built with these options, dnsmasq
   has no library dependencies other than libc */

#define HAVE_DHCP
#define HAVE_DHCP6
#define HAVE_TFTP
#define HAVE_SCRIPT
#define HAVE_AUTH
#define HAVE_IPSET
#define HAVE_LOOP
#define HAVE_DUMPFILE

/* Build options which require external libraries.

   Defining HAVE_<opt>_STATIC as _well_ as HAVE_<opt> will link the library statically.

   You can use "make COPTS=-DHAVE_<opt>" instead of editing these.
*/

/* #define HAVE_LUASCRIPT */
/* #define HAVE_DBUS */
/* #define HAVE_IDN */
/* #define HAVE_LIBIDN2 */
/* #define HAVE_CONNTRACK */
/* #define HAVE_CRYPTOHASH */
/* #define HAVE_DNSSEC */


/* Default locations for important system files. */

#ifndef LEASEFILE
#   if defined(__FreeBSD__) || defined (__OpenBSD__) || defined(__DragonFly__) || defined(__NetBSD__)
#      define LEASEFILE "/var/db/dnsmasq.leases"
#   elif defined(__sun__) || defined (__sun)
#      define LEASEFILE "/var/cache/dnsmasq.leases"
#   elif defined(__ANDROID__)
#      define LEASEFILE "/data/misc/dhcp/dnsmasq.leases"
#   else
#      define LEASEFILE "/var/lib/misc/dnsmasq.leases"
#   endif
#endif

#ifndef CONFFILE
#   if defined(__FreeBSD__)
#      define CONFFILE "/usr/local/etc/dnsmasq.conf"
#   else
#      define CONFFILE "/etc/dnsmasq.conf"
#   endif
#endif

#ifndef RESOLVFILE
#   if defined(__uClinux__)
#      define RESOLVFILE "/etc/config/resolv.conf"
#   else
#      define RESOLVFILE "/etc/resolv.conf"
#   endif
#endif

#ifndef RUNFILE
#   if defined(__ANDROID__)
#      define RUNFILE "/data/dnsmasq.pid"
#    else
#      define RUNFILE "/var/run/dnsmasq.pid"
#    endif
#endif

/* platform dependent options: these are determined automatically below

HAVE_LINUX_NETWORK
HAVE_BSD_NETWORK
HAVE_SOLARIS_NETWORK
   define exactly one of these to alter interaction with kernel networking.

HAVE_GETOPT_LONG
   defined when GNU-style getopt_long available.

HAVE_SOCKADDR_SA_LEN
   defined if struct sockaddr has sa_len field (*BSD)
*/

#if defined(__UCLIBC__)
#define HAVE_LINUX_NETWORK
#if defined(__UCLIBC_HAS_GNU_GETOPT__) || \
   ((__UCLIBC_MAJOR__==0) && (__UCLIBC_MINOR__==9) && (__UCLIBC_SUBLEVEL__<21))
#    define HAVE_GETOPT_LONG
#endif
#undef HAVE_SOCKADDR_SA_LEN
#if defined(__UCLIBC_HAS_IPV6__) && defined(USE_IPV6)
#  ifndef IPV6_V6ONLY
#    define IPV6_V6ONLY 26
#  endif
#elif !defined(NO_IPV6)
#  define NO_IPV6
#endif

/* This is for glibc 2.x */
#elif defined(__linux__)
#define HAVE_LINUX_NETWORK
#define HAVE_GETOPT_LONG
#undef HAVE_SOCKADDR_SA_LEN

#elif defined(__FreeBSD__) || \
      defined(__OpenBSD__) || \
      defined(__DragonFly__) || \
      defined(__FreeBSD_kernel__)
#define HAVE_BSD_NETWORK
/* Later versions of FreeBSD have getopt_long() */
#if defined(optional_argument) && defined(required_argument)
#   define HAVE_GETOPT_LONG
#endif
#define HAVE_SOCKADDR_SA_LEN

#elif defined(__APPLE__)
#define HAVE_BSD_NETWORK
#define HAVE_GETOPT_LONG
#define HAVE_SOCKADDR_SA_LEN
#define NO_IPSET
/* Define before sys/socket.h is included so we get socklen_t */
#define _BSD_SOCKLEN_T_
/* Select the RFC_3542 version of the IPv6 socket API.
   Define before netinet6/in6.h is included. */
#define __APPLE_USE_RFC_3542
/* Required for Mojave. */
#ifndef SOL_TCP
#  define SOL_TCP IPPROTO_TCP
#endif
#define NO_IPSET

#elif defined(__NetBSD__)
#define HAVE_BSD_NETWORK
#define HAVE_GETOPT_LONG
#define HAVE_SOCKADDR_SA_LEN

#elif defined(__sun) || defined(__sun__)
#define HAVE_SOLARIS_NETWORK
#define HAVE_GETOPT_LONG
#undef HAVE_SOCKADDR_SA_LEN
#define ETHER_ADDR_LEN 6

#endif

/* rules to implement compile-time option dependencies and
   the NO_XXX flags */

#ifdef NO_TFTP
#undef HAVE_TFTP
#endif

#ifdef NO_DHCP
#undef HAVE_DHCP
#undef HAVE_DHCP6
#endif

#if defined(NO_DHCP6)
#undef HAVE_DHCP6
#endif

/* DHCP6 needs DHCP too */
#ifdef HAVE_DHCP6
#define HAVE_DHCP
#endif

#if defined(NO_SCRIPT)
#undef HAVE_SCRIPT
#undef HAVE_LUASCRIPT
#endif

/* Must HAVE_SCRIPT to HAVE_LUASCRIPT */
#ifdef HAVE_LUASCRIPT
#define HAVE_SCRIPT
#endif

#ifdef NO_AUTH
#undef HAVE_AUTH
#endif

#if defined(NO_IPSET)
#undef HAVE_IPSET
#endif

#ifdef NO_LOOP
#undef HAVE_LOOP
#endif

#ifdef NO_DUMPFILE
#undef HAVE_DUMPFILE
#endif

#if defined (HAVE_LINUX_NETWORK) && !defined(NO_INOTIFY)
#define HAVE_INOTIFY
#endif

/* Define a string indicating which options are in use.
   DNSMASQ_COMPILE_OPTS is only defined in dnsmasq.c */

#ifdef DNSMASQ_COMPILE_OPTS

static char *compile_opts =
"IPv6 "
#ifndef HAVE_GETOPT_LONG
"no-"
#endif
"GNU-getopt "
#ifdef HAVE_BROKEN_RTC
"no-RTC "
#endif
#ifndef HAVE_DBUS
"no-"
#endif
"DBus "
#ifndef HAVE_UBUS
"no-"
#endif
"UBus "
#ifndef LOCALEDIR
"no-"
#endif
"i18n "
#if defined(HAVE_LIBIDN2)
"IDN2 "
#else
 #if !defined(HAVE_IDN)
"no-"
 #endif
"IDN "
#endif
#ifndef HAVE_DHCP
"no-"
#endif
"DHCP "
#if defined(HAVE_DHCP)
#  if !defined (HAVE_DHCP6)
     "no-"
#  endif
     "DHCPv6 "
#endif
#if !defined(HAVE_SCRIPT)
     "no-scripts "
#else
#  if !defined(HAVE_LUASCRIPT)
     "no-"
#  endif
     "Lua "
#endif
#ifndef HAVE_TFTP
"no-"
#endif
"TFTP "
#ifndef HAVE_CONNTRACK
"no-"
#endif
"conntrack "
#ifndef HAVE_IPSET
"no-"
#endif
"ipset "
#ifndef HAVE_AUTH
"no-"
#endif
"auth "
#if !defined(HAVE_CRYPTOHASH) && !defined(HAVE_DNSSEC)
"no-"
#endif
"cryptohash "
#ifndef HAVE_DNSSEC
"no-"
#endif
"DNSSEC "
#ifdef NO_ID
"no-ID "
#endif
#ifndef HAVE_LOOP
"no-"
#endif
"loop-detect "
#ifndef HAVE_INOTIFY
"no-"
#endif
"inotify "
#ifndef HAVE_DUMPFILE
"no-"
#endif
"dumpfile";

#endif /* defined(HAVE_DHCP) */
